### PR TITLE
added options for embeddings dimension reduction

### PR DESCRIPTION
**Pull request by nadyadevani3112**
provided two options of embeddings dimension reduction: UMAP or T-SNE

### CODE REVIEW

File: biotrainer/config/embedding_options.py [hunk 1]
```diff
@@ -144,5 +145,60 @@ def allow_download(self) -> bool:
...
+    def allow_multiple_values(self) -> bool:
+        return False
+
+    @property
+    def possible_values(self) -> List[Any]:
+        return ["umap", "tsne"]
+
+    @classproperty
+    def allowed_protocols(self) -> List[Protocol]:
+        return [Protocol.sequence_to_class, Protocol.sequence_to_value]
```
Comment (SebieF) [inline biotrainer/config/embedding_options.py:168]: You could use Protocol.using_per_sequence_embeddings() here instead.

File: biotrainer/config/embedding_options.py [hunk 2]
```diff
@@ -144,5 +145,60 @@ def allow_download(self) -> bool:
...
+    def allow_multiple_values(self) -> bool:
+        return False
+
+    @staticmethod
+    def _is_value_valid(config_option: ConfigOption, value) -> bool:
+        return type(value)==int and value > 0
+
+    @classproperty
+    def allowed_protocols(self) -> List[Protocol]:
+        return [Protocol.sequence_to_class, Protocol.sequence_to_value]
```
Comment (SebieF) [inline biotrainer/config/embedding_options.py:195]: Here as well (Protocol.using_per_sequence_embeddings())

File: biotrainer/config/embedding_options.py [hunk 3]
```diff
@@ -144,5 +145,60 @@ def allow_download(self) -> bool:
...
+    @classproperty
+    def required(self) -> bool:
+        return False
+
+
+class ReducedDimension(EmbeddingOption, ConfigOption):
+
+    @classproperty
+    def name(self) -> str:
+        return "reduced_dimension"
```
Comment (SebieF) [inline biotrainer/config/embedding_options.py:179]: I am thinking about whether "reduced_dimensions" or even "reduce_to_dimensions" would be indicating a bit clearer what is done by this config option, what do you think?

File: biotrainer/trainers/trainer.py [hunk 4]
```diff
@@ -173,6 +177,14 @@ def _create_and_load_embeddings(self) -> Dict[str, Any]:
 
         # Mapping from id to embeddings
         id2emb = embedding_service.load_embeddings(embeddings_file_path=embeddings_file)
+        if (self._dimension_reduction_method and 
+            self._reduced_dimension and
+            len(id2emb)>=3 and
+            list(id2emb.values())[0].shape[0]>=3):
```
Comment (SebieF) [inline biotrainer/trainers/trainer.py:183]: I would suggest moving this to a separate function to check the condition, and maybe also include another check for the protocol here.

File: pyproject.toml [hunk 5]
```diff
@@ -56,6 +56,8 @@ onnx = "1.16.1"
 onnxscript = "^0.1.0.dev20240806"
 onnxruntime = "^1.19.0"
 pandas = "^2.2.3"
+jupyter = "^1.1.1"
```
Comment (SebieF) [inline pyproject.toml:59]: Please remove the jupyter dependency as it is optional and leads to some dependency overhead that is not required for all users :)

File: biotrainer/embedders/embedding_service.py [hunk 6]
```diff
@@ -230,6 +232,42 @@ def _save_and_reset_embeddings(self, embeddings: Dict[str, ndarray], last_save_i
         del embeddings
         return last_save_id, {}
 
+    @staticmethod
+    def reduce_embeddings_dimension(
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:236]: The name might be confusing because per-protein embeddings are also referred to as "reduced". Maybe call it "def embeddings_dimensionality_reduction"?

File: biotrainer/embedders/embedding_service.py [hunk 7]
```diff
@@ -230,6 +232,42 @@ def _save_and_reset_embeddings(self, embeddings: Dict[str, ndarray], last_save_i
...
+        embeddings: Dict[str, Any],
+        dimension_reduction_method: str,
+        reduced_dimension: int) -> Dict[str, Any]:
+        """
+        Reduces the dimension of per-residue embeddings or per-protein embeddings
+
+        Parameters:
+            embeddings (Dict[str, ndarray]): Dictionary of embeddings.
+            embedder: The embedder used for reducing embeddings.
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:245]: Please add the missing function parameters here (and remove embedder)

File: biotrainer/embedders/embedding_service.py [hunk 8]
```diff
@@ -230,6 +232,42 @@ def _save_and_reset_embeddings(self, embeddings: Dict[str, ndarray], last_save_i
         del embeddings
         return last_save_id, {}
 
+    @staticmethod
+    def reduce_embeddings_dimension(
+        embeddings: Dict[str, Any],
+        dimension_reduction_method: str,
+        reduced_dimension: int) -> Dict[str, Any]:
+        """
+        Reduces the dimension of per-residue embeddings or per-protein embeddings
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:241]: It actually only works for per-protein embeddings, doesn't it?

File: biotrainer/embedders/embedding_service.py [hunk 9]
```diff
@@ -230,6 +232,42 @@ def _save_and_reset_embeddings(self, embeddings: Dict[str, ndarray], last_save_i
...
+
+        Parameters:
+            embeddings (Dict[str, ndarray]): Dictionary of embeddings.
+            embedder: The embedder used for reducing embeddings.
+
+        Returns:
+            out (Dict[str, ndarray]): Dictionary of reduced embeddings.
+        """
+        sorted_keys = sorted(list(embeddings.keys()))
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:250]: Please add a logging statement for start like: 
`logger.info(f"Starting embeddings dimensionality reduction via method {dimension_reduction_method}")`

File: biotrainer/embedders/embedding_service.py [hunk 10]
```diff
@@ -230,6 +232,42 @@ def _save_and_reset_embeddings(self, embeddings: Dict[str, ndarray], last_save_i
...
+            all_embeddings.shape[1]])
+        dimension_reduction_method_dict = {
+            "umap": UMAP(n_components=reduced_dimension),
+            "tsne": TSNE(
+                n_components=reduced_dimension,
+                perplexity=min(30, reduced_dimension))
+            }
+        embeddings_reduced_dimensions = dimension_reduction_method_dict[
+            dimension_reduction_method].fit_transform(all_embeddings)
+        return {sorted_keys[i]:torch.tensor(embeddings_reduced_dimensions[i]) for i in range(len(sorted_keys))}
```
Comment (SebieF) [inline biotrainer/embedders/embedding_service.py:268]: Please also add an end logging statement:
`logger.info(f"Finished embeddings dimensionality reduction!")`

**Review by SebieF — CHANGES_REQUESTED** [review]
Thank you so much for your contribution to biotrainer!! :) I provided some feedback on your changes, in general they look very good. 

I also rebased your branch to be consistent with the latest version of the develop branch, such that you do not have to do this somewhat tedious work of resolving conflicts: https://github.com/biocentral/biotrainer/commits/feature/emb-dim-redu-rebased/ I will provide you with detailed instructions how to use that before applying the suggestions in the next comment. Thanks alot again for your contribution!

**Comment by SebieF** [comment]
P.S.: Do not worry about the broken Windows tests, they are failing because of onnx at the moment: https://github.com/sacdallago/biotrainer/issues/111

File: biotrainer/config/embedding_options.py (hunk 1 above)
Comment (nadyadevani3112) [reply biotrainer/config/embedding_options.py:168]: I copied this part from the Interaction class in general_options.py, should I change it there too?

File: biotrainer/config/embedding_options.py (hunk 3 above)
Comment (nadyadevani3112) [reply biotrainer/config/embedding_options.py:179]: how about n_reduced_components?

File: biotrainer/embedders/embedding_service.py (hunk 8 above)
Comment (nadyadevani3112) [reply biotrainer/embedders/embedding_service.py:284]: yes sorry I wrote that before you said that it should only work for per-protein embeddings

File: biotrainer/embedders/embedding_service.py (hunk 7 above)
Comment (nadyadevani3112) [reply biotrainer/embedders/embedding_service.py:288]: sorry I copied this from the _reduce_embeddings function and forgot to change it

**Comment by nadyadevani3112** [comment]
i couldn't "git push --force-with-lease" successfully, had to pull and merge some conflicts before pushing it again successfully

**Comment by nadyadevani3112** [comment]
thank you for all your feedbacks, really appreciate them! I have made the changes, are you able to see the commits and review the PR again?

**Comment by nadyadevani3112** [comment]
hi sorry it took me a while to implement the changes. I have also updated the config file options documentation and added a new example on how to apply the dimension reduction.

**Review by SebieF — APPROVED** [review]

**Comment by SebieF** [comment]
Hi! Thanks for applying the changes and all your work! The final state looked good to me. In order to fix the conflicts with the current develop branch, I cherry picked your commits and made some tiny additions to them with you as the co-author. Then I had to force push the updated branch in order to be able to merge the PR. I hope that was okay for you! You can still find the original branch with the merges here: https://github.com/biocentral/biotrainer/tree/feature/embedding-dimension-reduction-keep

Thanks again for your contribution. If you plan to contribute to biotrainer in the future, feel free to reach out to me via my university mail address (see my [profile](https://github.com/SebieF)). I would also be curious to hear how you are using biotrainer for your application or research :)
